core/src/client/mod.rs

```diff
@@ -388,6 +474,27 @@ impl<Notif> Drop for Subscription<Notif> {
 	}
 }
 
+impl<Notif> Stream for Subscription<Notif>
```

Comment:
technically we should be able to do this instead but seems to some waker issue which doesn't always occur


```rust
impl<Notif> Stream for Subscription<Notif>
where
	Notif: DeserializeOwned,
{
	type Item = Result<Notif, SubscriptionError>;

	fn poll_next(
		self: std::pin::Pin<&mut Self>,
		cx: &mut std::task::Context<'_>,
	) -> std::task::Poll<Option<Self::Item>> {
		let mut this = Pin::new(self);

		let mut fut = this.rx.inner.recv();

		match fut.poll_unpin(cx) {
			Poll::Ready(Ok(v)) => Poll::Ready(Some(serde_json::from_value(v).map_err(Into::into))),
			Poll::Ready(Err(RecvError::Closed)) => Poll::Ready(None),
			Poll::Ready(Err(RecvError::Overflowed(n))) => Poll::Ready(Some(Err(SubscriptionError::Lagged(n)))),
			Poll::Pending => Poll::Pending,
		}
	}
}

```

Reply:
seems like some race-condition that doesn't occur with the default stream impl

Reply:
yeah ok I see the recv future has a `listener` which needs dropped when poll_next returns which makes it only work if channel is non-empty and it may not be woken up again after Poll::Pending

Reply:
aight, I see that I could add a field `pending` to the struct to be able to keep the state but tricky with borrowing....

Reply:
just to prove my point this hack fixes the stream impl and nothing else is causing that:

```rust
impl Stream for SubscriptionRx {
	type Item = Result<serde_json::Value, SubscriptionError>;

	fn poll_next(
		self: std::pin::Pin<&mut Self>,
		cx: &mut std::task::Context<'_>,
	) -> std::task::Poll<Option<Self::Item>> {
		let mut this = Pin::new(self);

		loop {
			if let Some(f) = this.wake.as_mut() {
				futures_util::ready!(f.poll_unpin(cx));
			}

			// The timer has expired and mark it as done;
			this.wake = None;

			// Attempt to receive a message.
			match this.inner.try_recv() {
				Ok(msg) => return Poll::Ready(Some(Ok(msg))),
				Err(TryRecvError::Closed) => return Poll::Ready(None),
				Err(TryRecvError::Overflowed(n)) => {
					return Poll::Ready(Some(Err(SubscriptionError::Lagged(n))));
				}
				Err(TryRecvError::Empty) => {
					tracing::info!("sub empty");
				}
			}

			if this.wake.is_none() {
				tracing::info!("enable sleep 3 seconds");
				this.wake = Some(futures_timer::Delay::new(Duration::from_secs(3)));
			}
		}
	}
}
```

Reply:
/cc @jsdw any ideas here? :)

Reply:
Mmm I see the issue! You can't store `this.inner.rx.recv()` because of the lifetime, but would need to otherwise dropping it throws the waker away and so nothing is woken up.

By the looks of it, `async_broadcast::Receiver` impls Stream, so you _could _ expose a way to call `self.inner.rx.inner.poll_next` directly instead?

Reply:
yeah I'm already doing that in this PR but the issue with that is https://docs.rs/async-broadcast/latest/src/async_broadcast/lib.rs.html#1385, thus the "lagged/overflowed" is ignored and that's not propagated on the stream.

The entire point of this manual impl is to include "lagged/overflowed" error as well....

Reply:
Thus, the behavior between `StreamExt::next` and `Subscription::recv` are different and for example in subxt we are relying on the stream impl and if the stream lags then we will replace old items and the user is not notified about it.

Not ideal and not the intention with this PR ^^

Reply:
Ooh I see; I didn't pay enough attention to notice that the vehaviours were different! That sucks. It doesn't look like you can recreate the logic in their `poll_next` and tweak it in the way you'd need offhand, so it feels like the only options left are:

- Open issue/PR with `async-broadcast` people to see if we can expose what we need there (perhaps by writing a Stream impl that returns all of the details exposed by `recv`, which needn't be the default or chaneg any of their APIs, or
- Write our own naive async broadcast channel so that we have total control over it, or
- Try and find something else out there that does what we want